framework/src/meshgenerators/MeshRepairGenerator.C

```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
+            else if (num_fixed_nodes == 10)
+              _console << "Node stitching will now proceed silently." << std::endl;
           }
         }
       }
     }
-    _console << "Number of overlapping nodes which got merged: " << _num_fixed_nodes << std::endl;
-    if (mesh->allow_renumbering())
-      mesh->renumber_nodes_and_elements();
-    else
-    {
-      mesh->remove_orphaned_nodes();
-      mesh->update_parallel_id_counts();
-    }
   }
+  _console << "Number of overlapping nodes which got merged: " << num_fixed_nodes << std::endl;
+  if (mesh->allow_renumbering())
+  {
+    mesh->remove_orphaned_nodes();
+    mesh->renumber_nodes_and_elements();
+  }
+  else
+  {
+    mesh->remove_orphaned_nodes();
+    mesh->update_parallel_id_counts();
+  }
+}
 
-  // Flip orientation of elements to keep positive volumes
-  if (_fix_element_orientation)
-    MeshTools::Modification::orient_elements(*mesh);
-
-  // Disambiguate any block that has elements of multiple types
-  if (_elem_type_separation)
+void
+MeshRepairGenerator::separateSubdomainsByElementType(std::unique_ptr<MeshBase> & mesh) const
+{
+  std::set<subdomain_id_type> ids;
+  mesh->subdomain_ids(ids);
+  // loop on sub-domain
+  for (auto & id : ids)
   {
-    std::set<subdomain_id_type> ids;
-    mesh->subdomain_ids(ids);
-    // loop on sub-domain
-    for (auto & id : ids)
+    // Gather all the element types and blocks
+    // ElemType defines an enum for geometric element types
+    std::set<ElemType> types;
+    // loop on elements within this sub-domain
+    for (auto & elem : mesh->active_subdomain_elements_ptr_range(id))
+      types.insert(elem->type());
+
+    if (types.size() > 1)
     {
-      // Gather all the element types and blocks
-      // ElemType defines an enum for geometric element types
-      std::set<ElemType> types;
-      // loop on elements within this sub-domain
-      for (auto & elem : mesh->active_subdomain_elements_ptr_range(id))
-        types.insert(elem->type());
-
-      if (types.size() > 1)
+      auto next_block_id = MooseMeshUtils::getNextFreeSubdomainID(*mesh);
+      subdomain_id_type i = 0;
+      for (const auto type : types)
       {
-        auto next_block_id = MooseMeshUtils::getNextFreeSubdomainID(*mesh);
-        subdomain_id_type i = 0;
-        for (const auto type : types)
-        {
-          auto new_id = next_block_id + i++;
-          // Create blocks when a block has multiple element types
-          mesh->subdomain_name(new_id) = mesh->subdomain_name(id) + "_" + Moose::stringify(type);
-
-          // Re-assign elements to the new blocks
-          for (auto elem : mesh->active_subdomain_elements_ptr_range(id))
-            if (elem->type() == type)
-              elem->subdomain_id() = new_id;
-        }
+        auto new_id = next_block_id + i++;
+        // Create blocks when a block has multiple element types
+        mesh->subdomain_name(new_id) = mesh->subdomain_name(id) + "_" + Moose::stringify(type);
```

Comment:
This is really nice.  IMHO we really ought to be doing this at the libMesh level, eventually.  The conflict between "ExodusII blocks can't be heterogeneous" and "ExodusII block ids are easy to lazily map 1-to-1 with libMesh subdomain ids" should *not* have resolved out to "libMesh subdomains can't be heterogeneous", but rather to "libMesh::ExodusII_IO should be less lazy".

Reply:
yeah that could also be the fix. I think there is some value in splitting the subdomains even outside of exodus IO lack of support. 

what if we introduce element-specific residual computation for example